File: Linked_Lists/python/palindrome.py
from LinkedList import  LinkedList
from Stack import Stack


def compare_lists(node1, node2):
    while node1.next is not None:
        if node1.data != node2.data:
            return False

        node1 = node1.next
        node2 = node2.next

    return True


def reverse_linked_list(node):
    if node.next is None:
        result_list = LinkedList()
        result_list.append(node.data)

        return result_list

    result_list = reverse_linked_list(node.next)
    result_list.append(node.data)

    return result_list


def is_palindrome(linked_list):
    reversed_list = reverse_linked_list(linked_list.head)

    return compare_lists(linked_list.head, reversed_list.head)


def is_palindrome2(linked_list):
    slow = linked_list.head
    fast = linked_list.head
    s = Stack()

    while fast is not None and fast.next is not None:
        s.push(slow.data)

        slow = slow.next
        fast = fast.next.next

    if fast is not None:
        slow = slow.next

    while slow is not None:
        top = s.pop()

        if slow.data != top:
            return False

        slow = slow.next

    return True


def main():
    linked_list = LinkedList()
    linked_list.append(1)
    linked_list.append(1)
    linked_list.append(2)
    linked_list.append(3)
    linked_list.append(2)
    linked_list.append(1)
    linked_list.append(2)

    print(is_palindrome2(linked_list))


if __name__ == '__main__':
    main()

File: Linked_Lists/go/palindrome.go
package main

import (
  "fmt"
  "types"
)


func compareLists(node1 *types.Node, node2 *types.Node) bool {
  for node1.Next != nil {
    if node1.Data != node2.Data {
      return false
    }

    node1 = node1.Next
    node2 = node2.Next
  }

  return true
}


func reverseLinkedList(node *types.Node) types.LinkedList {
  if node.Next == nil {
    var resultList types.LinkedList
    resultList.Append(node.Data)

    return resultList
  }

  resultList := reverseLinkedList(node.Next)
  resultList.Append(node.Data)

  return resultList
}


func isPalindrome(list types.LinkedList) bool {
  reversedList := reverseLinkedList(list.Head)

  return compareLists(list.Head, reversedList.Head)
}


func isPalindrome2(list types.LinkedList) bool {
  slow := list.Head
  fast := list.Head
  var s types.Stack

  for fast != nil && fast.Next != nil {
    s.Push(slow.Data)

    slow = slow.Next
    fast = fast.Next.Next
  }

  if fast != nil {
    slow = slow.Next
  }

  for slow != nil {
    top, _ := s.Pop()

    if slow.Data != top {
      return false
    }

    slow = slow.Next
  }

  return true
}


func main() {
  var list types.LinkedList

  list.Append(1)
  list.Append(2)
  list.Append(3)
  list.Append(1)
  list.Append(3)
  list.Append(2)
  list.Append(1)

  fmt.Println(isPalindrome2(list))

}
